pkg/api/store/nodetemplate/store.go

```diff
@@ -24,43 +24,95 @@ type nodeTemplateStore struct {
+	NodeTemplateClient v3.NodeTemplateInterface
 }
 
 func Wrap(store types.Store, npLister v3.NodePoolLister, secretLister corev1.SecretLister, ntClient v3.NodeTemplateInterface) types.Store {
-	s := &nodeTemplateStore{
-		Store:                 store,
+	t := &transformer{
+		NodeTemplateClient: ntClient,
+	}
+	s := &transform.Store{
+		Store:       store,
+		Transformer: t.filterLegacyTemplates,
+	}
+	return &nodeTemplateStore{
+		Store:                 s,
 		NodePoolLister:        npLister,
 		CloudCredentialLister: secretLister,
-		NodeTemplateClient:    ntClient,
-	}
-	return &transform.Store{
-		Store:       s,
-		Transformer: s.filterLegacyTemplates,
 	}
 }
 
-func (s *nodeTemplateStore) filterLegacyTemplates(apiContext *types.APIContext, schema *types.Schema, data map[string]interface{}, opt *types.QueryOptions) (map[string]interface{}, error) {
+func (t *transformer) filterLegacyTemplates(apiContext *types.APIContext, schema *types.Schema, data map[string]interface{}, opt *types.QueryOptions) (map[string]interface{}, error) {
 	ns, _ := data["namespaceId"].(string)
-	name, _ := data["name"].(string)
+	id, _ := data["id"].(string)
 	if ns != namespace.NodeTemplateGlobalNamespace {
-		s.NodeTemplateClient.Controller().Enqueue(ns, name)
+		// nodetemplate may not have been migrated yet, enqueue
+		t.NodeTemplateClient.Controller().Enqueue(ns, strings.TrimPrefix(id, ns+":"))
 		return nil, nil
 	}
 	return data, nil
 }
 
+func (s *nodeTemplateStore) ByID(apiContext *types.APIContext, schema *types.Schema, id string) (map[string]interface{}, error) {
+	originalNs, originalID := getNsAndID(id)
+	fullMigratedID := getMigratedID(originalNs, originalID)
+	migratedNs, migratedID := getNsAndID(fullMigratedID)
+	data, err := s.Store.ByID(apiContext, schema, fullMigratedID)
+	if err != nil {
+		return nil, replaceIDInError(err, migratedID, originalID, migratedNs, originalNs)
+	}
+
+	data["namespaceId"] = originalNs
+	data["id"] = id
+	return data, nil
+}
+
+func getMigratedID(ns, id string) string {
+	if ns == namespace.NodeTemplateGlobalNamespace {
+		return fmt.Sprintf("%s:%s", ns, id)
+	}
+
+	return fmt.Sprintf("%s:nt-%s-%s", namespace.NodeTemplateGlobalNamespace, ns, id)
+}
+
+func getNsAndID(id string) (string, string) {
+	idParts := strings.SplitN(id, ":", 2)
+	if len(idParts) == 1 {
+		return "", idParts[0]
+	}
+	return idParts[0], idParts[1]
+}
+
+func replaceIDInError(err error, id, newID, ns, newNs string) error {
+	if apiError, ok := err.(*httperror.APIError); ok {
+		msg := strings.Replace(apiError.Message, id, newID, -1)
+		msg = strings.Replace(msg, ns, newNs, -1)
+		err = httperror.NewAPIError(apiError.Code, msg)
+	}
+	return err
+}
+
 func (s *nodeTemplateStore) Delete(apiContext *types.APIContext, schema *types.Schema, id string) (map[string]interface{}, error) {
+	originalNs, originalID := getNsAndID(id)
+	fullMigratedID := getMigratedID(originalNs, originalID)
+	migratedNs, migratedID := getNsAndID(fullMigratedID)
```

Comment:
Can we move the three steps to a separate method and call that instead in all the Store API's?